File: HSPT2020/spiderman/spiderman.cpp
#include <bits/stdc++.h>
#define all(a) begin(a), end(a)
#define nl '\n'
typedef long long int ll;
int oo = -1^1<<31;
using namespace std;

bool bfs(vector<vector<int>> adjList, pair<int, int> badEdge) {
	queue<int> q;
	vector<bool> visited(adjList.size(), false);
	q.push(0);
	visited[0] = true;
	while (!q.empty()) {
		int u = q.front();
		q.pop();
		for (int v : adjList[u]) {
			if (visited[v] || (u == badEdge.first && v == badEdge.second)) continue;
			q.push(v);
			visited[v] = true;
		}
	}
	return visited[adjList.size()-1];
}

void solve() {
	int n, m; cin >> n >> m;
	vector<vector<int>> adjList(n);
	for (int i = 0; i < m; i++) {
		int u, v; cin >> u >> v;
		u--; v--;
		adjList[u].push_back(v);
		adjList[v].push_back(u);
	}
	pair<int, int> badEdge;
	for (int u = 0; u < n; u++) {
		for (int v : adjList[u]) {
			badEdge = {u, v};
			if (!bfs(adjList, badEdge)) {
				cout << "Halt, Spiderman! Plans Thwarted!" << nl;
				return;
			}
		}
	}
	cout << "How Sad, Perpetrator Triumphed" << nl;
}

int main()
{
    cin.tie(0)->sync_with_stdio(0);
	cin.exceptions(cin.failbit);

	int tt; cin >> tt;
	for (int t = 0; t < tt; t++)
		solve();

    return 0;
}

File: HSPT2020/spiderman/spiderman.py
def bfs(adjList, skipU, skipV, n):

    # Basic BFS with a skipped edge
    q = []
    visited = [False for _ in range(n)]
    q.append(0)
    visited[0] = True
    while q:
        u = q.pop(0)
        for v in adjList[u]:
            if visited[v]:
                continue
            if u == skipU and v == skipV:
                continue
            q.append(v)
            visited[v] = True
    return visited[n-1]

def solve():

    n, m = [int(x) for x in input().split()]

    # List comprehension for 2D array
    adjList = [[] for _ in range(n)]

    # For every edge
    for _ in range(m):
        # Get and 0 index the indices
        u, v = [int(x) for x in input().split()]
        u-=1
        v-=1

        # Create edge between u and v
        print("n = " + str(n) + " m = " + str(m))
        print("u = " + str(u) + " v = " + str(v))
        adjList[u].append(v)
        adjList[v].append(u)

    # For every node
    for u in range(n):
        # For every edge in that node
        for v in adjList[u]:
            skipU, skipV = u, v
            # BFS without edge, if it fails we win
            if not bfs(adjList, skipU, skipV, n):
                print("Halt, Spiderman! Plans Thwarted!")
                return

    # Made it through every edge safely
    print("How Sad, Perpetrator Triumphed")


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        solve()